Improve how libraries are located and various cleanup

tarsius opened the pull request:
This is the same as https://github.com/tromey/emacs-ffi/pull/21. Like that, this pr also includes the changes in https://github.com/tromey/emacs-ffi/pull/22.  Let me know whether you also want me to do multiple pull-requests.

I've improved one of the old commits ("Search for ffi-module on the load-path"):

- I'm not sure it is necessary to add the optional `nosuffix` and `path` arguments.  You might want to drop them and only add them we actually need them.  Currently this just mirrors `locate-library`, which might, or might not, be reason enough to add this here.

- It is necessary to load the module at compile-time as well.  Else the byte-compiler is very unhappy about all the unknown functions.

I've also added new commits:

- Consistently use spaces for indentation in `ffi.el`.

  That's the "new" convention for elisp.
  
  This causes conflicts with other prs, but I am going to resolve those in a bit.
  
  `ffi.c` should also be indented consistently.  I also recommend going with spaces.  If you decide to go with tabs instead, you would still have to make sure they are used consistently and you would have to set `tab-width` as a file-local variable, to ensure everyone uses the same setting.

- Addressed most of the compiler errors and warnings in `test.el`.

Some necessary things are not included:

- There are more serious issues the byte-compiler tells us about, but most of those are already addressed by https://github.com/tromey/emacs-ffi/pull/19.

- The C compiler pointed out that `config.h` has to be included, and doing that indeed makes the compilation succeed.  Someone else will have to figure out why that is, and document it in the commit message.

- You might want to start thinking about what Emacs releases you want to support and add CI to test on all of them.  You might want to do that before merging any prs.

nnicandro committed:
[WIP] Perform work in macro calling environment

In `define-ffi-function' and `ffi--struct-union-helper' the work to
define ffi functions (using `ffi--prep-cif' and `ffi--dlsym') and types
was done in the macro environment. This works fine during runtime, but
fails for complied code since the objects would be created at compile
time instead of at runtime. By moving the work to the macro's calling
environment we avoid this problem.

Co-authored-by: Jonas Bernoulli <jonas@bernoul.li>

tarsius committed:
emacs_module_init: Return early on subsequent invocations

Initializing the module a second time results in a segfault.

An attempt to initialize it a second time would occur when we evaluate
"ffi.el" a second time (e.g., if we `eval-buffer' after making changes
to some lisp code during development).

tarsius committed:
[KLUDGE] module: Include config.h

Without this we get many errors like this:

In file included from /.../emacs-module.h:29,
                 from ffi-module.c:16:
/.../lib/stddef.h:128:3: error: #error "Please include config.h first."
  128 |  #error "Please include config.h first."
      |   ^~~~~

All I know about this is that the compiler told me to do it and after
I did so, it stopped complaining.

tarsius commented:
> add CI to test on all of them.

I strongly suggest that you keep it simple. I.e., don't depend on Cask or anything else that brings in a pile of dependencies and complexity, without actually providing any (okay I concede, "only very little") value.

I *can* help with that, in case you have no or little experience with this, but I would prefer not to. (But feel free to ask me to have a look before it goes live.)

tarsius committed:
define-ffi-{library,function}: Use same name for variable and function

Spamming the global namespace with symbols like `G123' is not okay.

Also define the variables at run-time, using `defvar'.

suhail-singh commented:
> This is the same as https://github.com/tromey/emacs-ffi/pull/21.

Thank you!

> Let me know whether you also want me to do multiple pull-requests.

Single PR is fine at this stage, thanks.  In future, after the code has been
cleaned up a bit and we have CI I would prefer more focused PRs, but we're not
there yet.

> It is necessary to load the module at compile-time as well.

Is this something your PR already accomplishes (I haven't yet looked at the
code, and it'll take me a couple of days before I am able to)?  Or is this
something that is outstanding?

> Consistently use spaces for indentation in `ffi.el`.  ...  This causes
> conflicts with other prs, but I am going to resolve those in a bit.

Thank you.

> ffi.c` should also be indented consistently.  I also recommend going with
> spaces.

I agree.  Could you please add this as part of the PR as well?  Preferably all
whitespace changes in a single commit, but as long as whitespace changes aren't
mixed in with non-whitespace changes multiple commits would be okay as well.

> Addressed most of the compiler errors and warnings in `test.el`.

Thank you!  For my reference when reviewing, and since we don't yet have a
consistent dev environment set up, could you please share a snippet of the
invocations you ran as well the warnings and errors you observe.

> There are more serious issues the byte-compiler tells us about, but most of
> those are already addressed by https://github.com/tromey/emacs-ffi/pull/19.

If you do get around to rebasing that onto this, I would appreciate the help.
Unless, you believe that it would be better/easier if
https://github.com/tromey/emacs-ffi/pull/19 is merged prior to this PR.
Thoughts?

> The C compiler pointed out that `config.h` has to be included, and doing that
> indeed makes the compilation succeed.

Included how/where?  Could you share details, please.  It's been a while since
I've worked on a C project so there'll be a fair amount of (re-)learning on my
part here wrt the build process.

> You might want to start thinking about what Emacs releases you want to support
> and add CI to test on all of them.

Agreed, eventually.  Do you have recommendations on releases that ought to be
supported (other than 29.x onwards)?

> You might want to do that before merging any prs.

I won't yet be prioritizing CI work myself, but I would be open to reviewing
contributions on that if you or someone else decides to.

> I *can* help with that, in case you have no or little experience with this,
> but I would prefer not to.

A reference to a similar project (an Emacs module), if you're aware, with a
straightforward, but effective CI would help.  If you're not aware, please
ignore this request.

suhail-singh commented:
@tarsius I have incorporated the changes from your `harmless` branch.  Thank you
for separating those changes from the rest.  Could you please rebase on top of
current `master`?

Additionally, if not too much effort, please consider some tweaks based on below
in case they're not too much work for you.  They will help me when reviewing the
outstanding changes (which I have only briefly skimmed as of now).  In no
particular order:

1. It would help me if commits that are addressing shortcomings such as failures
   / bugs, came with some test case.  For instance, the commit to include an
   `initialized` flag.
2. It would help me if commits included a brief explanation of
   the motivation behind the proposed change.  For instance, the commit to
   `provide` the `ffi-module` feature.  While for this change I'm fairly
   confident as to the motivation, a few words (e.g. ensure `ffi-module` behaves
   similarly to other Emacs features) would help.  A good example of outlining
   the motivations was the commit where you set `indent-tabs-mode` - I really
   appreciated the clarity there.
3. For certain other changes, it would help to include some additional comments.
   For instance, it's not clear why we override the value of `load-suffixes` in
   `ffi--locate-module`.  Specifically, what's the failure mode, if any, that
   that override is trying to guard against?
4. Please add docstrings to any new macros or functions or existing macros or
   functions where you are making changes to the arglist.  In general, I would
   like docstrings on all functions and macros, but we can do this piecemeal.
5. Please add a copyright line for yourself in files you modify.

Finally, the last commit notes that it is fixing compiler errors and warnings.
Does that mean, on your machine, code checked out on that commit produces no
errors nor warnings?  If not, could you please note warnings and errors not yet
addressed?  I prefer to track these details in the git repository, as opposed to
discussions on `github`, `gitlab` etc.

tarsius commented:
I'll start by creating another branch of harmless changes. Probably later today.

tarsius commented:
> > Addressed most of the compiler errors and warnings in `test.el`.
> 
> Thank you! For my reference when reviewing, and since we don't yet have a consistent dev environment set up, could you please share a snippet of the invocations you ran as well the warnings and errors you observe.

It would be quite a bit of work for me to reproduce this warnings at this point, because I have already reordered the commits and because the Makefile is lacking.

So I can only give you this warning:

```
$ emacs -Q --batch -L . -f batch-byte-compile test.el

In toplevel form:
test.el:1:1: Warning: file has no ‘lexical-binding’ directive on its first line
```

After fixing that and some other issues, the byte-compiler would say something like.

```
The value of struct-type is unused.
```

I don't want to spend any time getting it to spit out the exact warning again. Instead I will work on setting up proper tooling for convenient byte-compilation.

suhail-singh commented on ffi-module.c:
```diff
@@ -13,12 +13,14 @@ GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with this.  If not, see <https://www.gnu.org/licenses/>.  */
 
+#include <config.h>
```
```elisp
#include <config.h>
```

On my machine, including this breaks `make all`.  Excluding it `make all`
succeeds.

For reference, I used the below for a quick build environment (before moving to
something more controlled):

```shell
guix shell gcc-toolchain coreutils-minimal binutils make libltdl
```

tarsius replied:
I'm also on Guix (though I have installed these packages in my home profile for the time being, i.e., even less controlled).

I'll leave it out from pull-requests for the time being.

suhail-singh commented on ffi.el:
```diff
@@ -28,35 +28,57 @@
+MODULE should be the file name of the module, a string, which must omit
+the suffix.  Search for the module in `load-path', unless optional PATH
+is specified, in which case search that list of directories instead."
+    ;; Override `load-suffixes' to avoid matching lisp libraries.  On
```
```elisp
    ;; Override `load-suffixes' to avoid matching lisp libraries.  On
```

Why is this necessary?

tarsius replied:
In the years I've been waiting for the license to be specified, I have forgotten. I am guessing that I ran into a situation where there actually was a module and library with matching names.

Note the other reason: it makes lookup a bit faster. I think that is reason enough to add this.

suhail-singh replied:
> it makes lookup a bit faster.

Don't `load` and others perform the lookup in order?  Have I misunderstood how
`load-suffixes` is used?

> I think that is reason enough to add this.

Performance tuning is something I would only want to consider after a CI and
some benchmarks are in place.  And even then, only when motivated by data.

If omitting this change breaks something, that's different.  However, in that
case please update the relevant code comment.

suhail-singh commented on ffi.el:
```diff
@@ -65,39 +87,39 @@
+             ,docstring)
+          result-forms)
     (cl-mapc
      (lambda (slot type offset)
```
```elisp
    (push `(defvar ,name (apply #',definer-function ',field-types)
```

What's the reason for inlining `(apply #',definer-function ',field-types)`?

tarsius replied:
I am not the author of that commit and I haven't studied it yet. 

This moves evaluation from macro-expansion time to run-time, which is what this commit is all about.

suhail-singh replied:
I see, thanks.

For this PR, I propose that we focus on the changes that you are sure about
which bring some immediate value.  Once we have those merged in, we can tackle
the others in subsequent PRs.

suhail-singh commented:
> I don't want to spend any time getting it to spit out the exact warning
> again.

Understandable.

> Instead I will work on setting up proper tooling for convenient
> byte-compilation.

Thank you.

tarsius committed:
Load our module at compile-time as well

We use functions from the module in "ffi.el",
so the byte-compiler needs to know about them.

tarsius committed:
Locate modules on the load-path

We have to do this for the libraries used by "ffi libraries"
as well as for our own "ffi-module.so".

Without doing it for the latter, we used to get:

  $ emacs -Q --batch -L . -f batch-byte-compile ffi.el

  In toplevel form:
  ffi.el:31:2: Error: Module could not be opened: \
  "ffi-module.so", "ffi-module.so: cannot open shared object file: \
  No such file or directory"

Note that in this case we are actually in the directory that contains
the module but never-the-less it doesn't work with a relative name.

tarsius committed:
ffi--struct-union-helper: Cosmetics

tarsius committed:
ffi--lay-out-struct: Cosmetics

tarsius committed:
define-ffi-function: Cosmetics

tarsius committed:
define-ffi-function: Name arguments consistently

Name them `arg<N>', where <N> is 1 initially and then is increased by 1
for every argument.  Previously `G<N>' but while <N> was also increased
for every argument, it started out at a random number, so the arguments
would get different names every time this is evaluated/compiled.

suhail-singh commented:
@tarsius I am seeing some of the refactoring you're doing.  Thank you for the
descriptive commit messages.  However, please note that I am not yet paying
close attention to the commits themselves.  Please let me know when it's
somewhat stabilized and you'd like me to have another look.

tarsius committed:
[WIP] Improve Makefile

tarsius committed:
[WIP] CI

tarsius committed:
[wip]

tarsius commented:
> note that I am not yet paying close attention to the commits themselves.

That is appropriate, they are currently very much in flux and until ci is fully setup(which I'll do in my fork first) even are likely to come with regressions.

Oh, and it may be a few days until I get back to this.

suhail-singh commented:
> Oh, and it may be a few days until I get back to this.

All good.  Please do tag me in a comment when you'd like me to have another
look.

Regarding the CI, for what it's worth (and I believe you're of the same
opinion), my preference would be to start with the simplest thing that allows us
to catch regressions and/or issues that users will (realistically) encounter.
One possibility may be to pick a very simple C-library (`foo-lib`), and use it
to test the intended functionality.  A guix shell invocation that runs the code
which tests the bindings with `foo-lib` would be sufficient to get us started.

tarsius commented:
You most likely won't get back to this before next month.

suhail-singh commented:
> You most likely won't get back to this before next month.

@tarsius is there a typo in the above?  Regardless, even for myself (should
there be commits you'd like me to review), it would be after the first week of
December.

tarsius commented:
Yesh :grin:

suhail-singh commented:
@tarsius for your awareness, I will be releasing a version for `emacs-ffi` in
early January 2025.  It'll coincide with me packaging `emacs-ffi` for Guix.
Please use this information to prioritize changes as you see fit.

That being said, I intend to release as frequently as is necessary to make
useful changes available to users.  So while there may be some utility in
prioritizing some changes over others, there'd be little reason to rush any
changes.

suhail-singh commented:
@tarsius in the interest of this PR not stagnating, do you think there would be
utility in splitting up some of these changes?  Are there any here that, in your
opinion, are ready to be merged?

tarsius commented:
I have worked on this some more today and made some progress. This also reminded me that I am actually interested in working on this and that it is a great learning opportunity. I had previously forgotten about that because I felt rushed -- it had just become one more thing to be done with as fast as possible. Rushing it hinders learning and I have already learned enough to know how little I know.

Sure, it has already been a few months since the licensing terms were finally specified and it became actually possible to use this project. But I do not think it is a good idea to rush anything, just because you itch to move forward.

Yes, I am still interested in this, but it has also has been five years since I proposed the first iteration of these changes and I currently have other priorities. I could drop those for a while and rush something here, but I would rather do it properly. That way I can learn something, do a better job, and have fun while doing it.

Initially it was very encouraging that you created an organization, but now I am under the impression that you do not know more about this than I do. It also feels like you have taken control, by creating an organization named "emacs-ffi", to make this fork more "official". I saw no problem with that initially, taking it as a sign that you were going to put in the work.

But now it feels like you are not doing much beside trying to get other people to do the work, while you maintain control. Of course there's also a positive side to not just discarding work others have already done and start over, but combined with increasing pressure to get it done and the increasing likelihood of something suboptiomal being merged soon, I feel like my only options are to power through now, or else part ways.

I have decided to part ways. I'll keep my own fork and work on it once I feel motivated and have a need for this functionality.

I wish you luck with your fork. I'll certainly keep an eye on it and maybe we can merge some time down the line.

suhail-singh commented:
> but now I am under the impression that you do not know more about this than I
> do.

I have never maintained otherwise.  Was there something I said that gave you the
wrong impression?  If so, I sincerely apologize.

> It also feels like you have taken control, by creating an organization named
> "emacs-ffi", to make this fork more "official".

The intent with the `emacs-ffi` organization was (and continues to be), in fact,
the opposite.  Since I intend to upstream this work in Guix, doing so under my
personal account might have led people to form the inaccurate impression
regarding the extent of my contributions.  I am simply a maintainer.  Keeping
the code under a neutral organization side-steps that issue and has the nice
property that it allows for others to also take on that role without causing
disruption.  Further, the choice to keep the code on Github was a conscious
decision (my personal preference would have probably been either Gitlab or
Codeberg) and it was done so to more clearly reflect the relation that this fork
has with `tromey`'s.

> But now it feels like you are not doing much beside trying to get other people
> to do the work, while you maintain control.

I am sorry that you feel that way, but if that is an honest reflection of your
thoughts I thank you for sharing your feedback.

For what it's worth, the goal has (and continues to be) to share the work done
by `tromey` (and the improvements made by others over the years) with the wider
community under an open-source license.

Reading between the lines, I am getting the sense that you no longer have an
objection to my merging of `jkitchin`'s PR
#4.  As such I shall do so shortly, and in the coming days package the work on
Guix in the hopes that it is upstreamed.

> I wish you luck with your fork.

Likewise!
